aoc2022/src/Day7.hs

```diff
@@ -0,0 +1,260 @@
+fileLineParser = do
+  fsize <- decimal <* space
+  fname <- fnameParser
+  pure . FileOutputLine $ File fname fsize
+
+outputLinesParser :: Parser [OutputLine]
+outputLinesParser =
+  someTill
+    ((dirLineParser <|> fileLineParser) <* eol)
+    (void (lookAhead "$ ") <|> eof)
+
+lsParser :: Parser Instruction
+lsParser =
+  dbg "LsParser" $ Ls <$> ("ls" *> eol *> outputLinesParser)
+
+instructionParser :: Parser Instruction
+instructionParser =
+  "$ " *> (cdParser <|> lsParser)
+
+puzzleParser :: Parser [Instruction]
+puzzleParser = some instructionParser
+
+instance Eq PuzzleState where
+  (PuzzleState tree1 _ spaceMap1) == (PuzzleState tree2 _ spaceMap2) =
+    tree1 == tree2 && spaceMap1 == spaceMap2
+
+instance Show PuzzleState where
+  show (PuzzleState {trees, spaceMap}) =
+    "Tree "
+      <> show trees
+      <> ";\n Space map: "
+      <> show spaceMap
+
+emptyFileTree :: FileTree
+emptyFileTree = FileTree RootDir [] Set.empty
+
+initialState :: IO PuzzleState
+initialState =
+  let initialWorkingDirectory = ofAs [RootDir]
+   in PuzzleState Map.empty <$> initialWorkingDirectory <*> mempty
+
+step :: PuzzleState -> Instruction -> IO PuzzleState
+step state (Cd RootDir) = ofAs [RootDir] <&> (\x -> state {workingDirectory = x})
+step state (Cd namedDir@(NamedDir _)) =
+  let workdir = workingDirectory state
+   in push namedDir workdir $> state
+step state@(PuzzleState {workingDirectory}) (Cd DotDot) =
+  pop workingDirectory $> state
+step state@(PuzzleState {workingDirectory, trees = initialTrees}) (Ls output) =
+  do
+    -- this is fine, because we always at least have a RootDir,
+    -- it just _looks_ horrifying
+    Just cwd <- top workingDirectory
+    normPath <- pack <$> key workingDirectory "/"
+    let prefix = normPath <> "/"
+    pure $
+      state
+        { trees =
+            initialTrees
+              <> Map.singleton
+                normPath
+                ( FileTree
+                    { files =
+                        ( \case
+                            FileOutputLine f -> [f]
+                            _ -> []
+                        )
+                          =<< output,
+                      subTree =
+                        ( \case
+                            DirectoryLine (NamedDir name) -> Set.singleton $ prefix <> name
+                            _ -> Set.empty
+                        )
+                          `foldMap` output,
+                      directory = cwd
+                    }
+                )
+        }
+
+buildState :: [Instruction] -> IO PuzzleState
+buildState instructions =
+  initialState >>= \state -> foldlM step state instructions
+
+calculateNodeSize ::
+  IORef (Map Text Integer) ->
+  Map Text FileTree ->
+  Text ->
+  IO Integer
+calculateNodeSize directorySizeCache directories nodeKey =
+  case Map.lookup nodeKey directories of
+    Just (FileTree {subTree, files}) ->
+      do
+        cache <- readIORef directorySizeCache
```

Comment:
I had `trace` calls in to confirm whether I was getting cache hits or misses since I didn't want to traverse the whole tree every time, and they confirmed this was working as I expected. For bonus points, I could try using [`STRef` instead](https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-STRef.html) but that didn't go great when I tried it for my `Stack` type